### PR TITLE
Set dismissible cards on domain-only upsell carousel

### PR DESCRIPTION
**Pull request by gius80**
## Changes proposed in this Pull Request

This PR add the option to dismiss domain only upsell carousel's cards for 1 month or 1 week.

It's part of the new domain-only flow (pcYYhz-ye-p2)

carousel-close

carousel-open

## Testing instructions

_Prerequisites: you need a domain-only subscriptions to test this PR_

- Build this branch locally or open the live Calypso link
- Navigate to to `/manage/domains`
- Click on the "Hide this" button on one of the carda and verify that it's removed.
- You can reactivate the card through the Developer console, using this endpoint `/me/domains/[:domainName]/notices/[:noticeType]/message`, where `noticeType` is `upsellCarouselHideCreateSiteCardUntil` or `upsellCarouselHideAddEmailCardUntil`

### CODE REVIEW

File: client/my-sites/domains/domain-management/list/domain-only-upsell-carousel.tsx [hunk 1]
```diff
@@ -24,94 +57,192 @@ const DomainOnlyUpsellCarousel = ( props: DomainOnlyUpsellCarouselProps ): JSX.E
...
+		};
+		const noticeValue = moment().add( mapReminderToMomentArgs[ reminder ], 'days' ).toISOString();
+		dispatchRecordTracksEvent( 'calypso_domain_only_upsell_card_hide_card', {
+			source_card_type: sourceCardType,
+			reminder,
+		} );
+		props.updateDomainNotice( domain.domain, noticeType, noticeValue );
+	};
+
+	const renderCard = ( {
```
Comment (leonardost) [inline client/my-sites/domains/domain-management/list/domain-only-upsell-carousel.tsx:118]: Love this refactoring 😄 a minor nitpick is that it could have been done in a separate PR, just to reduce the scope and size of each one 😉

File: client/my-sites/domains/domain-management/list/domain-only-upsell-carousel.tsx [hunk 2]
```diff
@@ -24,94 +57,192 @@ const DomainOnlyUpsellCarousel = ( props: DomainOnlyUpsellCarouselProps ): JSX.E
...
+	const hideCard = (
+		reminder: HideCardDuration,
+		noticeType: UpsellCardNoticeType,
+		sourceCardType: string
+	) => {
+		setHideSiteCardOptionsVisible( false );
+		setHideEmailCardOptionsVisible( false );
+		const mapReminderToMomentArgs = {
+			'1w': 7,
+			'1m': 30,
```
Comment (leonardost) [inline client/my-sites/domains/domain-management/list/domain-only-upsell-carousel.tsx]: I think using 30 here can be misleading because months have different numbers of days 🤔 one idea (in place of adding a condition to add either `days` or `months`) would be doing something like this:
```
const mapReminderToMomentArgs = {
	'1w': { weeks: 1 },
	'1m': { months; 1 },
};
```
and then just doing `moment().add( mapReminderToMomentArgs[ reminder ] )`. I saw this in Moment's docs but haven't tested it 😄

Comment (leonardost) [reply client/my-sites/domains/domain-management/list/domain-only-upsell-carousel.tsx]: Also, I think `1w` and `1m` should be made into constants, maybe `ONE_WEEK` and `ONE_MONTH`.

File: client/my-sites/domains/domain-management/list/types.tsx [hunk 3]
```diff
@@ -1,13 +1,30 @@
...
+	isRequestingDomainNotices: boolean;
+	isUpdatingDomainNotices: boolean;
 };
 
 export type DomainOnlyUpsellCarouselProps = DomainOnlyUpsellCarouselOwnProps &
-	DomainOnlyUpsellCarouselConnectedProps;
+	DomainOnlyUpsellCarouselConnectedProps &
+	DomainOnlyUpsellCarouselStateProps;
+
+export type HideCardDuration = '1w' | '1m';
```
Comment (leonardost) [inline client/my-sites/domains/domain-management/list/types.tsx]: I'm not sure if it's valid typescript, but I think we should transform `1w` and `1m` into constants, something like
```
export type HideCardDuration = ONE_WEEK | ONE_MONTH;
```
One o the reasons is that `m` in this context could mean 1 month or 1 minute, so it's a little ambiguous.

File: client/my-sites/domains/domain-management/list/domain-only-upsell-carousel.tsx [hunk 4]
```diff
@@ -24,94 +57,192 @@ const DomainOnlyUpsellCarousel = ( props: DomainOnlyUpsellCarouselProps ): JSX.E
...
 	);
 };
 
-export default connect( null, { dispatchRecordTracksEvent: recordTracksEvent } )(
-	DomainOnlyUpsellCarousel
-);
+const shouldHideCard = ( date: string ): boolean => {
+	const reminderEnd = moment( new Date( date ) );
+	return reminderEnd.isValid() && moment().isBefore( reminderEnd );
+};
```
Comment (leonardost) [inline client/my-sites/domains/domain-management/list/domain-only-upsell-carousel.tsx]: I think this also needs a condition for when `date` is null, as `new Date( null )` returns the Unix epoch (1970-01-01) which might have unintended consequences. I'm not sure, but doesn't `moment( date )` work, without needing to transform `date` into a `Date` object?

File: client/state/domains/notices/actions.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+		type: DOMAINS_NOTICES_REQUEST,
+		domainName,
+	} );
+
+	return wpcom.req.get( { path: `/me/domains/${ domainName }/notices` } ).then(
+		( { states } ) => {
+			dispatch( {
+				type: DOMAINS_NOTICES_REQUEST_SUCCESS,
+				domainName,
+				notices: states[ domainName ],
```
Comment (leonardost) [inline client/state/domains/notices/actions.js]: When there's still no state for a domain, `states` is null, so we should use `states?.[ domainName ]` here.

**Review by leonardost — CHANGES_REQUESTED** [review]
Thanks for this update @gius80! The code looks good, I just needed to tweak one line in `actions.js` because the carousel wasn't appearing due to my domain states being clean - but after that, it worked great 🙂 

There are some other minor observations in the code, but a more general one is that maybe we don't need to have everything in the global store here 🤔 for example, as this card is only used in one place in Calypso, `isRequestingDomainNotices` and `isUpdatingDomainNotices` could be in `DomainOnlyUpsellCarousel`'s local state instead of in the global store. I'm wondering if it would make sense to even not use redux here, as all data is only used locally 🤔 

I did exactly the same thing in #58426 😄 creating many actions and reducers, and Kamen pointed out that it could be simpler.

File: client/my-sites/domains/domain-management/list/domain-only-upsell-carousel.tsx (hunk 1 above)
Comment (gius80) [reply client/my-sites/domains/domain-management/list/domain-only-upsell-carousel.tsx:118]: Yes you're totally right, but the more I looked that code the more I wanted to clean it 😅

File: client/my-sites/domains/domain-management/list/types.tsx (hunk 3 above)
Comment (gius80) [reply client/my-sites/domains/domain-management/list/types.tsx]: I agree, it's pretty confusing. I converted it in `enum` construct.

```
export enum HideCardDuration {
	ONE_WEEK = 'ONE_WEEK',
	ONE_MONTH = 'ONE_MONTH',
}
```

I also converted `UpsellCardNoticeType` in enum.

File: client/my-sites/domains/domain-management/list/domain-only-upsell-carousel.tsx (hunk 4 above)
Comment (gius80) [reply client/my-sites/domains/domain-management/list/domain-only-upsell-carousel.tsx]: In case date is `null` it's probably not a problem (it acts like an expired reminder) but I refactored the code to improve readability.

I had to check `null` date because `moment(null)` return `null`

```
const shouldHideCard = ( date: string | null ): boolean => {
	if ( ! date ) return false;
	const reminderEnd = moment( date );
	return reminderEnd.isValid() && moment().isBefore( reminderEnd );
};

```

File: client/state/domains/notices/actions.js (hunk 5 above)
Comment (gius80) [reply client/state/domains/notices/actions.js]: Good catch! I didn't notice this because I had manually created the message in advance through api  😅

**Comment by gius80** [comment]
> There are some other minor observations in the code, but a more general one is that maybe we don't need to have everything in the global store here 🤔 for example, as this card is only used in one place in Calypso, `isRequestingDomainNotices` and `isUpdatingDomainNotices` could be in `DomainOnlyUpsellCarousel`'s local state instead of in the global store. I'm wondering if it would make sense to even not use redux here, as all data is only used locally 🤔

Thanks for the review @leonardost, your feedbacks are always very helpful and valuable 🙏

I had this doubt too before I started, but then I thought that messages/notices can be generic and used in other contexts too (the api is generic). Maybe I'm just overthinking things 😅

**Comment by leonardost** [comment]
@gius80 I know how you feel, these architectural decisions are hard, sometimes I tend to overthink too 😅 and there's no right or wrong. The way I'm trying to approach things these days is by doing the simplest one first and iterating on it later if needed 🙂 in this case, I'm leaning towards keeping the state local, but as you said that info might be used in other contexts eventually 🤔

**Comment by gius80** [comment]
> @gius80 I know how you feel, these architectural decisions are hard, sometimes I tend to overthink too 😅 and there's no right or wrong. The way I'm trying to approach things these days is by doing the simplest one first and iterating on it later if needed 🙂 in this case, I'm leaning towards keeping the state local, but as you said that info might be used in other contexts eventually 🤔

At the end I decided to move everything to local state, definitely better 😉

**Review by leonardost — APPROVED** [review]
Thanks for sticking with this @gius80, it looks great! 😄 I think it's much more simple and straightforward now that state is local 🙂 If we eventually need to use this info elsewhere, we can move that info to the global state 👍

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7151355

Thank you @gius80 for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7151355).
